src/openai_public.rs

```diff
@@ -25,92 +25,117 @@ const IM_SEP: &str = "<|im_sep|>";
-            (IM_START.to_string(), 100264),
-            (IM_END.to_string(), 100265),
-            (IM_SEP.to_string(), 100266),
-            (ENDOFPROMPT.to_string(), 100276),
-        ])
-    }
+  pub fn cl100k_im() -> Result<Encoding, EncodingFactoryError> {
+    EncodingFactory::cl100k_with_special_tokens(&[
+      // end of text actually doesn't appear to be supported by the latest models! you can try by sending it in in the completion model and counting tokens
+      // (ENDOFTEXT.to_string(), 100257),
+      (FIM_PREFIX.to_string(), 100258),
+      (FIM_MIDDLE.to_string(), 100259),
+      (FIM_SUFFIX.to_string(), 100260),
+      (IM_START.to_string(), 100264),
+      (IM_END.to_string(), 100265),
+      (IM_SEP.to_string(), 100266),
+      (ENDOFPROMPT.to_string(), 100276),
+    ])
+  }
 
-    pub fn cl100k_with_special_tokens(special_tokens: &[(String, usize)]) -> Result<Encoding, EncodingFactoryError> {
-        let mergeable_ranks = load_tiktoken_bpe(
-            include_bytes!("../data/cl100k_base.tiktoken"),
-            "223921b76ee99bde995b7ff738513eef100fb51d18c93597a113bcffe865b2a7",
-        )
-        .map_err(|_| EncodingFactoryError::FailedToLoadEncoding)?;
-        let special_tokens: HashMap<String, usize> = special_tokens 
-        .iter()
-        .cloned()
-        .collect();
-        Encoding::new(
+  pub fn cl100k_with_special_tokens(
+    special_tokens: &[(String, usize)],
+  ) -> Result<Encoding, EncodingFactoryError> {
+    let mergeable_ranks = load_tiktoken_bpe(
+      include_bytes!("../data/cl100k_base.tiktoken"),
+      "223921b76ee99bde995b7ff738513eef100fb51d18c93597a113bcffe865b2a7",
+    )
+    .map_err(|_| EncodingFactoryError::FailedToLoadEncoding)?;
+    let special_tokens: HashMap<String, usize> = special_tokens.iter().cloned().collect();
+    Encoding::new(
             "cl100k_base",
             r"(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+",
             mergeable_ranks,
             special_tokens,
             None,
         )
         .map_err(|e| EncodingFactoryError::UnableToCreateEncoding(e.to_string()))
-    }
+  }
+
+  pub fn o200k_base() -> Result<Encoding, EncodingFactoryError> {
+    let mergeable_ranks = load_tiktoken_bpe(
+      include_bytes!("../data/o200k_base.tiktoken"),
+      "446a9538cb6c348e3516120d7c08b09f57c36495e2acfffe59a5bf8b0cfb1a2d",
+    )
+    .map_err(|_| EncodingFactoryError::FailedToLoadEncoding)?;
+    let special_tokens: HashMap<String, usize> =
+      [(ENDOFTEXT.to_string(), 199999), (ENDOFPROMPT.to_string(), 200018)]
+        .iter()
+        .cloned()
+        .collect();
+
+    const PAT_STR: &str = concat!(
+      r"[^\r\n\p{L}\p{N}]?[\p{Lu}\p{Lt}\p{Lm}\p{Lo}\p{M}]*[\p{Ll}\p{Lm}\p{Lo}\p{M}]+(?:'s|'t|'re|'ve|'m|'ll|'d)?|",
+      r"[^\r\n\p{L}\p{N}]?[\p{Lu}\p{Lt}\p{Lm}\p{Lo}\p{M}]+[\p{Ll}\p{Lm}\p{Lo}\p{M}]*(?:'s|'t|'re|'ve|'m|'ll|'d)?|",
+      r"\p{N}{1,3}|",
```

Comment:
why is the pat str different?

i believe python just passes the raw pat string into rust.

Reply:
also, let's use the same syntax they use in python for less confusion (i.e. writing a list [] with a .join, unless there's a reason that that doesn't work)

Reply:
updated!